Test data to illustrate bug in GenotypeGVCFs.removeNonRefAlleles()

bbimber opened the pull request:
@davidbenjamin @ldgauthier: in #6263 you added --force-output-intervals to GenotypeGVCFs, which forces the tool to output variants based on a whitelist of sites.  I believe this exposed a pre-existing, not related bug.  GenotypeGVCFsEngine.removeNonRefAlleles() currently assumes the input has only one alternate allele.  If the gVCF has a site with 3 or more alleles, GenotypeGVCFsEngine.removeNonRefAlleles() isnt going to work as intended.  If any NON_REF is found, it *should* remove ALT allele header lines and return the new VC with NON_REF removed.  It currently only does this if "newAlleles.size() == 1", which I assume is a proxy for not having alternates.  That assumes the input had only 2 alleles, which isnt safe.  This PR includes a fix for this.

When I started investigating this I made a repro case (the attached VCF) and test case in GenotypeGVCFsIntegration test that uses --force-output-intervals to illustrate this.  Now that the actual problem is clearer, I could understand if you dont want to add more test data to GATK.  

I tried to write a unit test for removeNonRefAlleles(), but it didnt seem like it was going to be easy to make a new instance of GenotypeGVCFsEngine.

bbimber commented:
I just noticed #6405, which seems like it might have some similarity here?

davidbenjamin commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java:
```diff
@@ -562,6 +562,43 @@ public void testForceOutput() {
+
+        final List<VariantContext> actualVC = VariantContextTestUtils.getVariantContexts(output1);
+
+        // every site has output
```
Should this comment instead say "no site has output"?

bbimber replied:
ok

davidbenjamin commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java:
```diff
@@ -562,6 +562,43 @@ public void testForceOutput() {
         Assert.assertTrue(actualVC2.get(1).isPolymorphicInSamples());
     }
 
+    @Test
```
Add a comment that this checks that the non-ref is removed even for multiple alt alleles

bbimber replied:
ok

davidbenjamin commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java:
```diff
@@ -562,6 +562,43 @@ public void testForceOutput() {
+        final ArgumentsBuilder argsWithSpecificSites = new ArgumentsBuilder()
+                .addReference(b37Reference)
+                .addVCF(input)
+                .addArgument(GenotypeGVCFs.FORCE_OUTPUT_INTERVALS_NAME, "1:1048236-1048236")
```
Could you also force output at the site 1051053, where there is a * allele?

bbimber replied:
ok

davidbenjamin commented on src/test/resources/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs/ForceOutputNonRef.g.vcf:
```diff
@@ -0,0 +1,142 @@
+##contig=<ID=Y,length=59373566,assembly=b37>
+##contig=<ID=MT,length=16569,assembly=b37>
+##source=SelectVariants
+#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	BCM-NEPRC-34593	BCM-NEPRC-34594	BCM-NEPRC-34595	BCM-NEPRC-34596	BCM-NEPRC-34597	BCM-NEPRC-34598	BCM-NEPRC-34599	BCM-NEPRC-34600	BCM-NEPRC-34601	BCM-NEPRC-34602
```
Could you doctor this VCF to include a site with two non-symbolic alt alleles in addition to the NON-REF?

bbimber replied:
ok

davidbenjamin commented:
> I just noticed #6405, which seems like it might have some similarity here?

That's a different bug.

davidbenjamin commented:
@bbimber Thank you for finding and fixing this!  Once you address the comments and add a unit test as I outlined above this should be ready to merge.

bbimber commented:
@davidbenjamin I think all requested changes are in, with the exception of modifying the gVCF to create a site with two ALT alleles in addition to NON_REF.  I'll try to find an example in our data, but I'm not sure I understand gVCF format enough to do that.  do you happen to know if you have existing test data multi-ALT gVCFs?

bbimber commented:
@davidbenjamin - nevermind.  gVCF now includes a site with two ALTs, plus NON_REF.

davidbenjamin commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java:
```diff
@@ -562,6 +564,60 @@ public void testForceOutput() {
+        // build intervals
+        final File output2 = createTempFile("output", ".vcf");
+        final File intervalFile = createTempFile("testForceOutputNonRef", ".intervals");
+        try (PrintWriter writer = new PrintWriter(IOUtil.openFileForWriting(intervalFile))) {
```
Instead of creating a file, just specify multiple interval literals
```
. . .
   .addArgument(GenotypeGVCFs.FORCE_OUTPUT_INTERVALS_NAME, "1:1048236-1048236")
   .addArgument(GenotypeGVCFs.FORCE_OUTPUT_INTERVALS_NAME, "1:1051053-1051053")
   .addArgument(GenotypeGVCFs.FORCE_OUTPUT_INTERVALS_NAME, "1:2364622-2364622")
   . . .
```

bbimber replied:
ok

davidbenjamin commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java:
```diff
@@ -562,6 +564,60 @@ public void testForceOutput() {
+        }
+
+        // No sites should be output
+        final ArgumentsBuilder argsWithAllSites = new ArgumentsBuilder()
```
`argsWithAllSites` is confusing because one would expect more output than for `argsWithSpecificSites`.  Therefore `argsWith(out)ForceCalling` would be clearer.

bbimber replied:
sure

bbimber commented:
One additional question: as I look more closely at the output, should GenotypeGVCFs prune unused alleles?  This I think this applies mostly in the situation where it outputs all sites, or where --force-output-sites is used.

davidbenjamin commented:
Yes, GenotypeGVCFs prunes alleles that lack sufficient evidence.  The force output intervals only guarantees that the site is emitted, but it may be emitted as hom ref.

bbimber commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsEngine.java:
```diff
@@ -185,7 +185,7 @@ private VariantContext regenotypeVC(final VariantContext originalVC, final Refer
      * @return variant context with the NON-REF alleles removed if multiallelic or replaced with NO-CALL alleles if biallelic
      */
-    private VariantContext removeNonRefAlleles(final VariantContext vc) {
+    protected static VariantContext removeNonRefAlleles(final VariantContext vc, Set<String> infoHeaderAltAllelesLineNames) {
```
ok

bbimber commented:
> Yes, GenotypeGVCFs prunes alleles that lack sufficient evidence. The force output intervals only guarantees that the site is emitted, but it may be emitted as monomorphic.

OK, but I think my test shows that's not happening.  In the --force-output-intervals case of testForceOutputNonRef(), all VCs in actualVC2 are not polymorphic.  Despite this, some of them have alternate alleles?  see the new test case I added at GenotypeGVCFsIntegrationTest line 614

bbimber commented:
@davidbenjamin I looked into this more.  From what I can tell, when a site is polymorphic, GenotypeGVCFs is going to trim to found alleles.  Under normal conditions, when a site is not polymorphic, it doesnt get output.  When output is forced (--force-output-interverals or output-all-sites), then the VC is just output as-is.  This PR initially fixed the bug in removeNonRefAlleles().  I think GenotypeGVCFs should also prune unused alleles.  If the site is not polymorphic, by definition I think that means removing all ALTs?  I just checked in a few tweaks:

1) I switched from the original test gVCF data I added to use an existing gVCF from GATK.  This has the advantage of containing actual callable sites, not adding new data, and contains existing multi-allelic sites.  

2) I added code to GenotypeGVCFs so unused alleles will be trimmed as part of removeNonRefAlleles.  I could understand if you want to change the name of that method now.  This is currently always true; however, I could see a rationale for making --retain-unused-alleles-from-force-output-sites as a command line argument (i'd argue to default to false).

bbimber commented:
@davidbenjamin sorry to bug, but we're really hoping to get this resolved.  see my last post - while i think it's clear GATK wasnt pruning alleles in any kind of force-output/output-all case, my latest changes add this and also add test cases around it.  i'm now using one of the pre-existing gVCFs as test data, which improves the test coverage as well.

bbimber commented:
@davidbenjamin again, sorry to keep bugging on this thread, but it's been a while and we're really hoping to push these changes through since they're blocking a project.  I believe I addressed everything in your review.  I did identify another (arguably pre-existing) issue in GenotypGVCFs that would be exposed whenever it runs in all-sites mode or in force-output mode - the lack of allele trimming.  This PR addresses that, including test cases, though I havent heard back about this.

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsEngine.java:
```diff
@@ -183,21 +185,37 @@ private VariantContext regenotypeVC(final VariantContext originalVC, final Refer
         // If NON_REF is the only alt allele, ignore this site
         final List<Allele> newAlleles = new ArrayList<>();
         // Only keep alleles that are not NON-REF
+        boolean altAllelesChanged = false;
```
Instead of tracking this boolean, why not replace `if(altAllelesChanged)` with `if(newAlleles.size() < vc.getNAlleles()`?

bbimber replied:
yes, good point

davidbenjamin commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsEngine.java:
```diff
@@ -183,21 +185,37 @@ private VariantContext regenotypeVC(final VariantContext originalVC, final Refer
+            else {
+                altAllelesChanged = true;
+            }
+        }
```
Can we simplify the bulk of this method to just
```
final Set<Allele> allelesToKeep = removeUnusedAlternates ? vc.getGenotypes().stream().flatMap(g->g.getAlleles().stream()).collect(Collectors.toSet()) : vc.getAlleles();
allelesToKeep.remove(Allele.NON_REF_ALLELE);
final List<Allele> newAlleles = vc.getAlleles().stream().filter(allelesToKeep::contains).collect(Collectors.toList());
```

bbimber replied:
I modified this slightly, but yes something like this is simpler than what's there.  The main difference is to ensure REF is retained

davidbenjamin commented:
> I added code to GenotypeGVCFs so unused alleles will be trimmed as part of removeNonRefAlleles. I could understand if you want to change the name of that method now. 

It would be good to change it to 'removeNonRefAndUnusedAltAlleles`.

> This is currently always true; however, I could see a rationale for making --retain-unused-alleles-from-force-output-sites as a command line argument (i'd argue to default to false).

I don't think we need to worry about such an argument.

davidbenjamin commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java:
```diff
@@ -562,6 +562,70 @@ public void testForceOutput() {
+
+        // No sites should be output
+        final File output1 = createTempFile("output", ".vcf");
+        final ArgumentsBuilder argsWithAllSites = new ArgumentsBuilder()
```
Looks like there's one last place to rename `argsWithAllSites`

bbimber replied:
ok

davidbenjamin commented:
I'm stepping into the 7 remaining test failures in the debugger, and I think they may just be fixed by overwriting the actual output of `GenotypeGVCFsIntegrationTest.testEntireVariant` onto the expected output for all the forced output test cases.  However, this exercise has revealed some other changes:

`GenotypeGVCFsEngine.callRegion` contains the line `final VariantContext mergedVC = merger.merge(variantsToProcess, loc, ref.getBase(), !outputNonVariants, false)`.  Since we want to remove the non-ref allele regardless, we should replace `!outputNonVariants` with `true`.  It is important to do this here, because only then does `regenotypeVC` correctly restrict the `AD` to the emitted alleles.  Also, doing this strips the variant QUAL from the output if it's monomorphic, which is desired (the `RGQ` field is in the output, so there remains a sense of the quality of the call).

Once you do that, could you check whether `removeNonRefAndUnusedAltAlleles` is still necessary?  In the integration tests, at least, I didn't see any alt alleles output except those called in a genotype.

bbimber commented:
I'll check this out.  I dont see local test failures, so i'm not sure what 7 you're referring to, but I agree callRegion() is worth checking out.  I am initially seeing this causing some other GenotypeGVCFsIntegrationTest test failures, but need to look more closely.

davidbenjamin commented:
@bbimber I'm not sure if you're able to see the Travis results but the tests here: https://travis-ci.com/broadinstitute/gatk/jobs/283620338 are failing.  I reproduced them locally while running `GenotypeGVCFsIntegrationTest.testEntireVariant`.

bbimber commented:
@davidbenjamin OK, i see them now.  these were not failing for me locally, so i'm going to need to look more.  I am fairly certain these didnt occur prior to my last commit (the errors about allele sorting are odd).  nonetheless your point about callRegion() might also be right.

davidbenjamin commented:
The error message about allele sorting is unhelpful -- it's from an integration test check for exact match of output vcf against an expected vcf and the "sort order" error really just means there are fewer alleles in the output than expected.  Since this change is what we want, the solution is just to overwrite the expected VCF.  I checked all the discordant files and found nothing wrong, so this is okay.

I also looked at all the output after the change `!outputNonVariants` --> `true` that I suggested, and it is definitely more correct.

bbimber commented:
there is one thing i dont like - if i revert the change i made in removeNonRefAndUnusedAltAlleles(), the one to simplify, these tests pass locally for me, with the test files as-is.  do you see this?

davidbenjamin commented:
Yes, that's because removing the extra alleles makes your output different from the "expected" VCF, causing the exact match check to fail.  It's okay; the expected VCF is wrong.  This is not surprising since non-default arguments don't always receive the same attention when people write tests.

bbimber commented:
If there is a non-ref site, but the sample has genotype calls (0/0), do you expect QUAL to be nulled out and AD to be removed?

bbimber commented:
This reworks callRegion() as you suggest and removes most other changes originally here.  Expected VCFs are updated and it includes the tests I added.  I rebased due to a conflict in gvcf.basepairResolution.includeNonVariantSites.vcf.  I think the changes in master are merged correctly, but dont know the context around those that well.

If you look at the diff, you'll see some changes in includeLowQualSites.vcf.  The version on this branch is updated from the GenotypeGVCFs output - do you think the dropping of the LowQual filter is right?

bbimber commented:
@davidbenjamin there is one test that failed.  is this possible an intermittent /timeout problem?  i dont have permission to restart them, but i dont see an actual test failure in it: https://travis-ci.com/broadinstitute/gatk/jobs/283688682

davidbenjamin commented on src/test/resources/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs/expected/includeLowQualSites.vcf:
```diff
@@ -38,14 +38,14 @@
+20	10012731	.	C	.	.	.	DP=4	GT:DP:RGQ	0/0:2:6	./.:0:0	0/0:2:6
+20	10012732	.	A	.	.	.	DP=4	GT:DP:RGQ	0/0:2:6	./.:0:0	0/0:2:6
+20	10012733	.	G	.	.	.	DP=4	GT:DP:RGQ	0/0:2:6	./.:0:0	0/0:2:6
+20	10012734	.	G	.	0.02	LowQual	DP=4;ExcessHet=3.01;MLEAC=.;MLEAF=.;MQ=NaN;RAW_MQ=7200.00	GT:DP:RGQ	0/0:2:6	0/0:0:0	0/0:2:6
```
It's weird that this monomorphic site still has a qual, but given that it was already doing thatI don't consider it your responsibility.  I will look into it later.

davidbenjamin commented:
@bbimber everything looks good!  One Travis job was failing with a transient error, so I restated it.  Since all the other jobs passed I'm pretty sure this will be ready to merge in an hour or two.
